File: tests/typesafety/python/higher_order_misuse.py
# expect: fails
# title: The wrapped fetch still rejects swapped arguments
# misuse-of: higher_order
from collections.abc import Callable


def retry[**P, R](times: int, operation: Callable[P, R]) -> Callable[P, R]:
    def attempt(*args: P.args, **kwargs: P.kwargs) -> R:
        failures = 0
        while True:
            try:
                return operation(*args, **kwargs)
            except ConnectionError:
                failures += 1
                if failures >= times:
                    raise
    return attempt


def fetch(url: str, timeout: int) -> str:
    return f"GET {url} within {timeout}s"


patient_fetch = retry(3, fetch)
result = patient_fetch(10, "https://example.test")

File: tests/typesafety/typescript/higher_order_misuse.ts
// expect: fails

class ConnectionLost extends Error {}

function retry<A extends unknown[], R>(
  times: number,
  operation: (...args: A) => R,
): (...args: A) => R {
  return (...args: A): R => {
    let failures = 0;
    for (;;) {
      try {
        return operation(...args);
      } catch (error) {
        if (!(error instanceof ConnectionLost)) throw error;
        failures += 1;
        if (failures >= times) throw error;
      }
    }
  };
}

function fetchPage(url: string, timeout: number): string {
  return `GET ${url} within ${timeout}s`;
}

const patientFetch = retry(3, fetchPage);
export const result = patientFetch(10, "https://example.test");
